library/std/src/sys_common/ucs2.rs

```diff
@@ -0,0 +1,177 @@
+//! Implementating of UCS-2 encoding. This is a lesser known subset of UTF-16 without surrogate
+//! pairs. It is used in UEFI
+#![allow(dead_code)]
+
+use crate::fmt;
+use crate::iter::FusedIterator;
+use crate::marker::PhantomData;
+use crate::num::NonZeroU16;
+use crate::ptr::NonNull;
+use crate::str;
+
+/// A struct to represent UCS-2 character
+#[unstable(feature = "ucs2", issue = "none")]
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+pub struct Ucs2Char {
+    value: NonZeroU16,
+}
+
+/// Format the code point as `U+` followed by four to six hexadecimal digits.
+/// Example: `U+1F4A9`
+#[unstable(feature = "ucs2", issue = "none")]
+impl fmt::Debug for Ucs2Char {
+    #[inline]
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(formatter, "U+{:04X}", self.value)
+    }
+}
+
+impl Ucs2Char {
+    #[unstable(feature = "ucs2", issue = "none")]
+    pub const REPLACEMENT_CHARACTER: Ucs2Char =
+        Ucs2Char { value: NonZeroU16::new(0xfffdu16).unwrap() };
+
+    pub(crate) const CR: Self = Ucs2Char { value: NonZeroU16::new(0x000du16).unwrap() };
+    pub(crate) const LF: Self = Ucs2Char { value: NonZeroU16::new(0x000au16).unwrap() };
+
+    fn new(value: NonZeroU16) -> Self {
+        Self { value }
+    }
+
+    #[unstable(feature = "ucs2", issue = "none")]
+    #[inline]
+    pub fn from_u16(c: u16) -> Option<Self> {
+        match NonZeroU16::try_from(c) {
+            Ok(x) => Some(Ucs2Char::new(x)),
+            Err(_) => None,
+        }
+    }
```

Comment:
This is problematic because the code points from U+D800 to U+DFFF are not valid scalar values.
```suggestion
    pub fn from_u16(c: u16) -> Option<Self> {
        match c {
            c => NonZeroU16::new(c),
            0xd800..=0xdfff => None,
        }
    }
```

Reply:
The match arms should be flipped. Otherwise the `None` arm is unreachable.

Reply:
Nice catch. I had thought about doing this in the future, but it had completely slipped my mind.